integration/kotlinx-coroutines-guava/src/ListenableFuture.kt

```diff
@@ -271,57 +278,46 @@ private class ToContinuation<T>(
+ * the [CoroutineExceptionHandler]. Cancelling a [Deferred] places that [Deferred] in the
+ * cancelling/cancelled states defined by [Job], which _can_ show the error. It's assumed that
+ * the [Deferred] pointing to the task will be used to observe any error outcome occurring after
+ * cancellation.
+ *
+ * This may be counterintuitive, but it maintains the error and cancellation contracts of both
+ * the [Deferred] and [ListenableFuture] types, while permitting both kinds of promise to point
+ * to the same running task.
  */
 private class ListenableFutureCoroutine<T>(
-    context: CoroutineContext,
-    private val future: SettableFuture<T>
-) : AbstractCoroutine<T>(context), Runnable  {
+    context: CoroutineContext
+) : AbstractCoroutine<T>(context) {
 
-    /**
-     * When registered as a [ListenableFuture] listener, cancels the returned [Coroutine] if
-     * [future] is successfully cancelled. By documented contract, a [Future] has been cancelled if
-     * and only if its `isCancelled()` method returns true.
-     *
-     * Any error that occurs after successfully cancelling a [ListenableFuture]
-     * created by submitting the returned object as a [Runnable] to an `Executor` will be passed
-     * to the [CoroutineExceptionHandler] from the context. The contract of [Future] does not permit
-     * it to return an error after it is successfully cancelled.
-     *
-     * By calling [asListenableFuture] on a [Deferred], any error that occurs after successfully
-     * cancelling the [ListenableFuture] representation of the [Deferred] will _not_ be passed to
-     * the [CoroutineExceptionHandler]. Cancelling a [Deferred] places that [Deferred] in the
-     * cancelling/cancelled states defined by [Job], which _can_ show the error. It's assumed that
-     * the [Deferred] pointing to the task will be used to observe any error outcome occurring after
-     * cancellation.
-     *
-     * This may be counterintuitive, but it maintains the error and cancellation contracts of both
-     * the [Deferred] and [ListenableFuture] types, while permitting both kinds of promise to point
-     * to the same running task.
-     */
-    override fun run() {
-        if (future.isCancelled) {
-            cancel()
-        }
-    }
+    // JobListenableFuture propagates external cancellation to `this` coroutine. See JobListenableFuture.
+    val future = JobListenableFuture<T>(this)
 
     override fun onCompleted(value: T) {
-        future.set(value)
+        future.complete(value)
     }
 
-    // TODO: This doesn't actually cancel the Future. There doesn't seem to be bidi cancellation?
     override fun onCancelled(cause: Throwable, handled: Boolean) {
-        if (!future.setException(cause) && !handled) {
-            // prevents loss of exception that was not handled by parent & could not be set to SettableFuture
+        if (!future.completeExceptionally(cause) && !handled) {
```

Comment:
It seems like (a long-existing) problem here: if `future {}` is cancelled **normally** (e.g. its parent got cancelled) and meanwhile someone completes the future from the outside, then `JobCancellationException` got reported into an exception handler, while we guarantee that no `CancellationException` will ever reach it.

This problem seems to exist in most of our `AbstractCoroutine` implementors, unfortunately. It would be nice if you could write a test that reproduces the problem and a fix for that (see: `MonoCoroutine.onCancelled`), I'll take care of the rest of the builders

Reply:
You are indeed right, I've checked the wrong source